crypto/dilithium/README.md

```diff
@@ -0,0 +1,71 @@
+
+### **Supported Versions**
+AWS-LC supports only Dilithium3 algorithm at this point. The NID assigned to Dilithium3 is `NID_DILITHIUM3` and the corresponding `PKEY` identifier is `EVP_PKEY_DILITHIUM3`.
+
+### **Source Code Origin and Modifications**
+The source code was taken from the primary source of Diltihium at [link](https://github.com/pq-crystals/dilithium), at [commit](https://github.com/pq-crystals/dilithium/commit/3e9b9f1412f6c7435dbeb4e10692ea58f181ee51) as of 9th May 2022.
+The `api.h`, `fips202.h` and `params.h` header files were modified [in this PR](https://github.com/awslabs/aws-lc/pull/655) to support our [prefixed symbols build](https://github.com/awslabs/aws-lc/blob/main/BUILDING.md#building-with-prefixed-symbols).
+
+Only the reference C implementation of Dilithium3 is currently integrated. The code is in the `pqcrystals-dilithium_dilithium3_ref` folder. The following changes were made to the code.
+
+* `randombytes.{h|c}` are deleted because we are using the randomness generation functions provided by AWS-LC.
+* `rng.{h|c}` are deleted because we are using the randomness generation functions provided by AWS-LC.
+* `symmetric-aes.c` are removed because we are using only the SHA3 based Dilithium.
+* `sign.c`: calls to `randombytes` function is replaced with calls to `pq_custom_randombytes` and the appropriate header file is included (`crypto/rand_extra/pq_custom_randombytes.h`).
+* `sign.c` was modified from `*mlen = -1;` to `*mlen = 0;` within the `crypto_sign_open` function. This was modified as `*mlen = -1;` is attempting to assign `-1` to an unsigned integer, giving the warning: ` warning C4245: '=': conversion from 'int' to 'std::size_t'`. This warning would cause Windows x86 builds to fail. See [pq-crystals/dilithium#65](https://github.com/pq-crystals/dilithium/issues/65) for upstream issue raised.
+* `sign.c` was modified to correct the documentation around `crypto_sign_verify`. See [pq-crystals/dilithium/#66](https://github.com/pq-crystals/dilithium/pull/66) for upstream PR.
+* `config.h`: `#define DILITHIUM_MODE 3`  is explicitly defined (to specify Dilithium3).
+* `reduce.c`: modify documentation comment for `reduce32`, fixing upper bound for return value `r`.
+* `aes256ctr.{h|c}`: these files were removed.
+* `config.h`: remove references to AES-mode dilithium
+* `symmetric.h`: remove references to AES-mode dilithium
+
+## **Example Usage** 
+The API is defined and documented in `include/openssl/evp.h`. To see examples of how to use Dilithium3, see `crypto/dilithium/p_dilithium_test.cc`. TLDR:
+
+```
+// Creates a new PKEY and PKEY context with Dilithium3
+EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_DILITHIUM3, nullptr);
+EVP_PKEY *dilithium_pkey = EVP_PKEY_new();
+
+// Generates a (public, private) key pair
+EVP_PKEY_keygen(ctx, &dilithium_pkey)
+
+// Sign
+bssl::ScopedEVP_MD_CTX md_ctx;
```

Comment:
- Nit: the example is mixing C and C++ style of using aws-lc. It'd be nice to settle on one (I'd prefer C).
- It would be good to add a full example of how to use the API (similarly to https://github.com/aws/aws-lc/tree/main/crypto/kem#full-example).

Reply:
I'm going to post a separate PR that will be solely concerned with dilithium documentation. we can continue this discussion there.